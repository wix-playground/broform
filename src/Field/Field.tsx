import * as React from 'react';
import {toJS, computed} from 'mobx';
import {inject, observer} from 'mobx-react';
import {FormController, FormField, FormMeta} from '../FormController';

export type ValidationFunction =
  | ((value: any, values?: any) => any | null)
  | ((value: any, values?: any) => Promise<any | null>);

export type FieldAdapter = ((adapterProps: AdapterProps) => JSX.Element) | React.ComponentClass<any> | React.SFC<any>;

export interface AdapterMetaInfo {
  errors: any | null;
  isDirty: boolean;
  isTouched: boolean;
  isActive: boolean;
  isValidating: boolean;
  custom: {[key: string]: any};
  isRegistered: boolean;
  form: FormMeta;
}

export interface AdapterProps {
  broform?: {
    name: string;
    meta: AdapterMetaInfo;
    value: any;
    setCustomState: (key: string, value: any) => void;
    setFormCustomState: (key: string, value: any) => void;
    onChange: (value: any) => void;
    onFocus: () => void;
    onBlur: () => void;
    validate: () => void;
  };
}

export type FormatterFunction = (value: any) => any;
export type EqualityCheckFunction = (newValue: any, oldValue: any) => boolean;

export interface InjectedFieldProps {
  controller?: FormController;
}

export interface OwnFieldProps {
  name: string;
  children?: (injectedAdapterProps: AdapterProps) => JSX.Element;
  adapter?: FieldAdapter;
  defaultValue?: any;
  onValidate?: ValidationFunction;
  onFormat?: FormatterFunction;
  onEqualityCheck?: EqualityCheckFunction;
  persist?: boolean;
  adapterProps?: any;
}

export interface FieldProps extends InjectedFieldProps, OwnFieldProps {}

@inject('controller')
@observer
export class Field extends React.Component<FieldProps> {
  static defaultProps = {
    isEqual: (newValue: any, oldValue: any) => newValue === oldValue,
    persist: false,
  };

  //meta info passed to Adapter
  @computed
  protected get meta(): AdapterMetaInfo {
    const controller = this.props.controller!;
    const {meta, errors} = this.field;

    const adapterErrors = toJS(errors);
    const custom = toJS(meta.custom);

    return {
      custom,
      errors: adapterErrors,
      isActive: meta.isActive,
      isDirty: meta.isDirty,
      isTouched: meta.isTouched,
      isValidating: meta.isValidating,
      isRegistered: meta.isRegistered,
      form: {
        isSubmitting: controller.isSubmitting,
        isValidating: controller.isValidating,
        isValid: controller.isValid,
        isDirty: controller.isDirty,
        isTouched: controller.isTouched,
        submitCount: controller.submitCount,
      },
    };
  }

  //field value, passed to adapter
  @computed
  protected get value(): any {
    return toJS(this.field.value, {
      detectCycles: false,
    });
  }

  @computed
  get field(): FormField {
    return this.props.controller!.fields.get(this.props.name) as FormField;
  }

  //custom state for field, passed to adapter
  protected setCustomState = (key: string, value: any) => {
    this.props.controller!.setFieldCustomState(this.props.name, key, value);
  };

  //focus handler, passed to adapter
  protected onFocus = (): void => {
    this.props.controller!.changeFieldActiveState(this.props.name, true);
  };

  //blur handler, passed to adapter
  protected onBlur = (): void => {
    this.props.controller!.changeFieldActiveState(this.props.name, false);
  };

  //value change handler, passed to adapter
  protected onChange = (value: any): void => {
    this.props.controller!.changeFieldValue(this.props.name, value);
  };

  //registers Field in FormController
  componentDidMount() {
    this.props.controller!.registerField(this, this.props);
  }

  //unregisters Field in FormController
  componentWillUnmount() {
    this.props.controller!.unRegisterField(this.props.name);
  }

  //render the adapter passed as `adapter` prop  with optional `adapterProps` prop,
  //or as children render function, broform prop is injected either way, but `adapterProps` are not passed in second case.
  render() {
    if (!this.field) {
      return null;
    }

    const {name} = this.props;
    const controller = this.props.controller!;

    const injectedAdapterProps: AdapterProps = {
      broform: {
        name,
        meta: this.meta,
        value: this.value,
        onChange: this.onChange,
        setCustomState: this.setCustomState,
        onFocus: this.onFocus,
        setFormCustomState: controller.setFormCustomState,
        onBlur: this.onBlur,
        validate: controller.validate,
      },
    };

    return this.props.adapter ? (
      <this.props.adapter {...injectedAdapterProps} {...this.props.adapterProps} />
    ) : (
      this.props.children!(injectedAdapterProps)
    );
  }
}
