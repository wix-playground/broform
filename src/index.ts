export {
  ValidationFunction,
  FieldAdapter,
  AdapterProps,
  AdapterMetaInfo,
  EqualityFunction,
  InjectedFieldProps,
  FieldProps,
  Field,
} from './Field';
export {FormProps, Form} from './Form';
export {
  FormController,
  FormAPI,
  FormValidationErrors,
  FormValues,
  FieldValidationError,
  FormField,
  FormFieldMeta,
} from './FormController';
export {FormPart} from './FormPart';
