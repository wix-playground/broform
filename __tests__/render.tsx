import * as React from 'react';
import {mount} from 'enzyme';
import {TestForm} from '../test/components/TestForm';
import {createTestFormDriver} from '../test/components/TestForm.driver';
import {InputAdapter} from '../test/components/InputAdapter';
import {Field} from '../src/Field';

test('Render', () => {
  const wrapper = mount(
    <TestForm>
      <Field name={TestForm.FIELD_ONE_NAME} adapter={InputAdapter} />
      <Field name={TestForm.FIELD_TWO_NAME}>
        {(props) => {
          return <InputAdapter {...props} />;
        }}
      </Field>
    </TestForm>,
  );
  const formDriver = createTestFormDriver({wrapper});

  expect(formDriver.get.serialized()).toMatchSnapshot();
});
