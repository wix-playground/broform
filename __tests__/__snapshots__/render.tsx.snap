// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1`] = `
<form
  data-hook="test-form"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    data-hook="field_one_name"
  >
    <input
      data-hook="input-field_one_name"
      name="field_one_name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value=""
    />
    <div
      data-hook="meta-props"
    >
      <span
        data-hook="meta_errors"
      >
        null
      </span>
      <span
        data-hook="meta_isActive"
      >
        false
      </span>
      <span
        data-hook="meta_isDirty"
      >
        false
      </span>
      <span
        data-hook="meta_isTouched"
      >
        false
      </span>
      <span
        data-hook="meta_isValidating"
      >
        false
      </span>
      <span
        data-hook="meta_isRegistered"
      >
        true
      </span>
      <span
        data-hook="meta_form:isSubmitting"
      >
        false
      </span>
      <span
        data-hook="meta_form:isValidating"
      >
        false
      </span>
      <span
        data-hook="meta_form:isValid"
      >
        true
      </span>
      <span
        data-hook="meta_form:isDirty"
      >
        false
      </span>
      <span
        data-hook="meta_form:isTouched"
      >
        false
      </span>
      <span
        data-hook="meta_form:submitCount"
      >
        0
      </span>
    </div>
    <span
      data-hook="set-custom-state"
      onClick={[Function]}
    />
    <span
      data-hook="validate"
      onClick={[Function]}
    />
  </div>
  <div
    data-hook="field_two_name"
  >
    <input
      data-hook="input-field_two_name"
      name="field_two_name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value=""
    />
    <div
      data-hook="meta-props"
    >
      <span
        data-hook="meta_errors"
      >
        null
      </span>
      <span
        data-hook="meta_isActive"
      >
        false
      </span>
      <span
        data-hook="meta_isDirty"
      >
        false
      </span>
      <span
        data-hook="meta_isTouched"
      >
        false
      </span>
      <span
        data-hook="meta_isValidating"
      >
        false
      </span>
      <span
        data-hook="meta_isRegistered"
      >
        true
      </span>
      <span
        data-hook="meta_form:isSubmitting"
      >
        false
      </span>
      <span
        data-hook="meta_form:isValidating"
      >
        false
      </span>
      <span
        data-hook="meta_form:isValid"
      >
        true
      </span>
      <span
        data-hook="meta_form:isDirty"
      >
        false
      </span>
      <span
        data-hook="meta_form:isTouched"
      >
        false
      </span>
      <span
        data-hook="meta_form:submitCount"
      >
        0
      </span>
    </div>
    <span
      data-hook="set-custom-state"
      onClick={[Function]}
    />
    <span
      data-hook="validate"
      onClick={[Function]}
    />
  </div>
  <div
    data-hook="form-values"
  >
    {}
  </div>
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
